### ISSUE TITLE
Support for Account hold, restore, pause, resubscribe 

### ISSUE THREAD

**Issue by rkotzy** [RevenueCat/purchases-android#171]
Google announced that starting November 1st, 2020 certain features will change from optional to either mandatory or on by default. You can read the full announcement here: https://android-developers.googleblog.com/2020/06/new-features-to-acquire-and-retain-subscribers.html

Currently, RevenueCat's entitlement system supports these states, but the dashboard, charts, and integrations do not. This means that you can enable those features for your app and it will work as expected within the app, but result in some inconsistencies in reporting and integrations (like webhooks). 

We do plan on fully supporting these features soon. This is a server-side change for RevenueCat, but we're creating the issue here so other Android developers can discover and follow this issue.

**Comment by ubuntudroid** [comment]
Any updates here?

**Comment by rkotzy** [comment]
We're going to be specing this out soon and should be deployed before the end of September. I don't think they'll be any required SDK update here and would just be RevenueCat server updates. Once the spec is finalized, will post any planned changes here that may be important (e.g. new properties in webhooks).

**Comment by ubuntudroid** [comment]
Great news! However, shouldn't there be some kind of SDK changes, e.g. a function to find out if a user is currently on a paused subscription or, more importantly, whether account hold is active? With the current entitlements we can just find out whether an entitlement is active or not.

**Comment by mguellsegarra** [comment]
Hi,
Any news about this?
Thank you!

**Comment by rkotzy** [comment]
Update here:

The RevenueCat backend is all setup to handle these new Google states now. This means that entitlement granting and price calculations will work as expected - so nothing will break.

1. If you're using webhooks, there will be some new/updated events coming soon:
    - `SUBSCRIPTION_PAUSED`: New event type to detect a paused subscription on your server. There is no "resumed" event, will just be the next renewal.
    - `BILLING_ISSUE`: Event will be updated to contain a flag to determine if the user has a grace period. You don't need to react to this flag, the "expiration_at" will still contain the correct grace period expiration date.

2. There will not be a specific event for the "On Hold" state. This is essentially a billing issue + the absence of a renewal.

3. There won't be any required SDK changes/updates.

Some comments have mentioned detecting these states from the SDK. 
- "Account Hold" - would be 30 days after the expiration date if this is enabled (fwiw Apple has always had this state and it lasts for 60 days)
- "Grace Period" - If the Entitlement is active and `billingIssueDetectedAt` is not null then subscriber is in a grace period.
- "Subscription Paused" - This state won't be able to be directly detected through the SDK currently, but is already logged as a future feature. Google recommends providing a deep link for the user to manage their subscription. This can be accessed through the `managementURL` property on the Purchaser Info object and shown to users with an expired subscription.
- "Payment Pending" - Not directly related to this change, but this state can be detected at the time of purchase through the PAYMENT_PENDING error: https://docs.revenuecat.com/docs/errors#--payment_pending.

**Comment by mguellsegarra** [comment]
Thank you Ryan for the update, and thank you all RevenueCat members for giving us this awesome SDK. Cheers!

**Comment by ubuntudroid** [comment]
Just FYI: https://medium.com/androiddevelopers/preparing-your-apps-for-the-latest-features-in-google-plays-billing-system-210ed5e50eaa

**Comment by Neo76289** [comment]
I just tried a subscription pause and it doesn't work. Each time it returns to me the entitlement exist and it is active. Google Docs says "**While the subscription is paused, the user doesn't have access to the subscription.** " [Google docs link](https://developer.android.com/google/play/billing/subs#pause). How do I know that a subscription is in pause status and lock premium content?

**Comment by lonewolfaka9** [comment]
+1

**Comment by vegaro** [comment]
From the docs:

> A subscription pause takes effect only after the current billing period ends. While the subscription is paused, the user doesn't have access to the subscription. 

The subscription pause **takes effect** at the end of the billing period. After the end of the billing period, it will be considered as paused and the user won't have access to the subscription. Before the end of the billing period is not considered as paused so the user should still have access. It is basically like a cancellation with an automatic restore after some period of time. This is the current behavior in RevenueCat.

**Comment by selimbenmrad** [comment]
Actually, I would highly appreciate to know how to implement the changes using the **react native sdk.**

I do not found this discussion helpful.

I think creating a post on the revenue cat blog might be very helpful for many other developer.

**Comment by rgusso** [comment]
> Actually, I would highly appreciate to know how to implement the changes using the **react native sdk.**
> 
> I do not found this discussion helpful.
> 
> I strongly believe creating a post on the revenue cat blog might be very helpful for many other developers.

and flutter

**Comment by Neo76289** [comment]
> From the docs:
> 
> > A subscription pause takes effect only after the current billing period ends. While the subscription is paused, the user doesn't have access to the subscription.
> 
> The subscription pause **takes effect** at the end of the billing period. After the end of the billing period, it will be considered as paused and the user won't have access to the subscription. Before the end of the billing period is not considered as paused so the user should still have access. It is basically like a cancellation with an automatic restore after some period of time. This is the current behavior in RevenueCat.


I tried again and it worked, the first time I didn't take into account that the "pause"starts from the end of the subscription.
Thanks for the explanation.

**Comment by rkotzy** [comment]
Going to mark this as closed now since everything has been released.
